### PR TITLE
UserData methods moved from Visual to Node

### PR DESCRIPTION
**Pull request by ahcorde**
# 🦟 Bug fix

Fixes https://github.com/ignitionrobotics/ign-rendering/issues/58

## Summary

This PR is part of the effort to consolidate `ign-gui` and `ign-gazebo` (related to this conversation https://github.com/ignitionrobotics/ign-gui/pull/231#discussion_r662702693) and it also targets this issue https://github.com/ignitionrobotics/ign-rendering/issues/58. 

The idea is to move the `UserData` and `SetUserData` methods to the Node class. I'm not sure how to proceed about the change in the API if I should tick-tock the methods in the Visual class, this change should not affect the API because `Node` is a parent class of `Visual`, it will break ABI but this PR is targeting `main`.

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: include/ignition/rendering/Node.hh [hunk 1]
```diff
@@ -302,6 +308,29 @@ namespace ignition
       /// \brief Remove all child nodes from this node
       /// This detaches all the child nodes but does not destroy them
       public: virtual void RemoveChildren() = 0;
+
+      /// \brief A map of custom key value data
+      protected: std::map<std::string, Variant> userData;
```
Comment (chapulina) [inline include/ignition/rendering/Node.hh:313]: This class doesn't have any other protected members yet. I wonder if we should make this private and PIMPLize the class instead. What do you think, @iche033 ?

I'm also not sure if it's best to put this in `BaseNode`, like it was in `BaseVisual`.

File: src/base/BaseScene.cc [hunk 2]
```diff
@@ -755,6 +755,7 @@ CameraPtr BaseScene::CreateCamera(const std::string &_name)
 CameraPtr BaseScene::CreateCamera(unsigned int _id, const std::string &_name)
 {
   CameraPtr camera = this->CreateCameraImpl(_id, _name);
+  camera->SetUserData("user-camera", false);
```
Comment (chapulina) [inline src/base/BaseScene.cc:758]: I think all cameras are created using this function, even camera sensors. I'd expect this to be something that is set after the camera is created by the caller of the function.

File: include/ignition/rendering/Node.hh [hunk 3]
```diff
@@ -302,6 +308,29 @@ namespace ignition
...
+      /// \param[in] _key Unique key
+      /// \param[in] _value Value in any type
+      public: virtual void SetUserData(const std::string &_key, Variant _value)
+      {
+        this->userData[_key] = _value;
+      }
+
+      /// \brief Get custom data stored in this visual
+      /// \param[in] _key Unique key
+      /// \param[in] _value Value in any type
```
Comment (chapulina) [inline include/ignition/rendering/Node.hh:325]:
```suggestion
      /// \return Value in any type
```

File: include/ignition/rendering/Node.hh (hunk 1 above)
Comment (iche033) [reply include/ignition/rendering/Node.hh:313]: hmm let's put it in `BaseNode`. That's a templated class so probably won't be able to pimpl'ize it.

Comment (ahcorde) [reply include/ignition/rendering/Node.hh]: done https://github.com/ignitionrobotics/ign-rendering/pull/358/commits/92c7ca15a10d05981fc22eb7a8a31a0e95ab95a2

File: include/ignition/rendering/Node.hh (hunk 3 above)
Comment (ahcorde) [reply include/ignition/rendering/Node.hh]: Done https://github.com/ignitionrobotics/ign-rendering/pull/358/commits/92c7ca15a10d05981fc22eb7a8a31a0e95ab95a2

**Comment by iche033** [comment]
there are a couple of clang warnings:
https://build.osrfoundation.org/job/ignition_rendering-ci-pr_any-homebrew-amd64/1604/clang/new/

**Comment by ahcorde** [comment]
@osrf-jenkins retest this please

**Comment by ahcorde** [comment]
@iche033 Fixed https://github.com/ignitionrobotics/ign-rendering/pull/358/commits/978b6bcef224364d7cc6eac106ad9e6094c4e310

**Review by iche033 — APPROVED** [review]

**Comment by iche033** [comment]
https://github.com/ignitionrobotics/ign-rendering/pull/361 should fix the github CI failing builds

**Review by chapulina — APPROVED** [review]

**Pull request merged**